### PR TITLE
[FLINK-27678] Support append-only table for file store.

### PR DESCRIPTION
**Pull request by openinx**
Add append-only table support for file store.

### CODE REVIEW

File: flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/source/FileStoreSourceSplitReader.java [hunk 1]
```diff
@@ -203,10 +209,21 @@ private class FileStoreRecordIterator implements BulkFormat.RecordIterator<RowDa
                         .orElse(null);
 
         private FileStoreRecordIterator() {
-            this.rowDataSupplier =
-                    valueCountMode
-                            ? new ValueCountRowDataSupplier(this::nextKeyValue)
-                            : new PrimaryKeyRowDataSupplier(this::nextKeyValue);
+            switch (writeMode) {
+                case APPEND_ONLY:
+                    this.rowDataSupplier = new AppendOnlyRowDataSupplier(this::nextKeyValue);
```
Comment (openinx) [inline flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/source/FileStoreSourceSplitReader.java:214]: Looks like we could just reuse the `PrimaryKeyRowDataSupplier` although its key is always a dummy empty key for an append only table. Let me revert this changes.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/AppendOnlyRowDataSupplier.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.store.file.KeyValue;
+import org.apache.flink.table.store.file.ValueKind;
+import org.apache.flink.util.Preconditions;
+
+import java.util.function.Supplier;
+
+/** A {@link Supplier} to read {@link RowData} from {@link KeyValue} for an append-only table. */
+public class AppendOnlyRowDataSupplier implements Supplier<RowData> {
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/AppendOnlyRowDataSupplier.java:30]: Remove this class.

**Review by JingsongLi — COMMENTED** [review]
Thanks for the contribution! Looks very nice~

I have two main comments:
- What is the strategy for assigning buckets to record, randomly?
- FileStoreSourceSplitGenerator needs break bucket to multiply splits per file.

**Comment by openinx** [comment]
> What is the strategy for assigning buckets to record, randomly?

For append-only table,  the assignment policy is determined by the SinkRecordConverter#bucket,  which will just calculate the hash code of the whole row to get the correct bucket.  I think it it the correct direction to go.

> FileStoreSourceSplitGenerator needs break bucket to multiply splits per file.

Yeah,  this is a great question which I did not address in this PR currently now.  I've just filed a separate issue to address it so that we can add separate unit tests for it.  Pls see: https://issues.apache.org/jira/browse/FLINK-27696

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/FileStoreImpl.java [hunk 3]
```diff
@@ -179,6 +184,25 @@ public RowType partitionType() {
...
+        return new FileStoreImpl(
+                tablePath,
+                schemaId,
+                options,
+                WriteMode.APPEND_ONLY,
+                user,
+                partitionType,
+                RowType.of(),
+                rowType,
+                null);
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/FileStoreImpl.java:203]: Add nullable annotation to all `MergeFunction`s.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/WriteMode.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+package org.apache.flink.table.store.file;
+
+/** Defines the write mode for flink table store. */
+public enum WriteMode {
+    APPEND_ONLY(
+            "append-only",
+            "The table can only accept append-only insert operations. All rows will be "
+                    + "inserted into the table store without any deduplication or primary/unique key constraint."),
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/WriteMode.java:26]: here is `or` or `and`?

File: flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/TableStoreFactoryOptions.java [hunk 5]
```diff
@@ -57,6 +59,27 @@ public class TableStoreFactoryOptions {
...
+                    .enumType(WriteMode.class)
+                    .defaultValue(WriteMode.CHANGE_LOG)
+                    .withDescription(
+                            Description.builder()
+                                    .text("Specify the write mode for table.")
+                                    .linebreak()
+                                    .text(
+                                            String.format(
+                                                    "%s: %s",
+                                                    WriteMode.APPEND_ONLY,
```
Comment (JingsongLi) [inline flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/TableStoreFactoryOptions.java:73]: Can you just use `DescribedEnum` just like `LogOptions.LogStartupMode`?

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreReadImpl.java [hunk 6]
```diff
@@ -83,6 +89,34 @@ public FileStoreRead withValueProjection(int[][] projectedFields) {
     @Override
     public RecordReader createReader(BinaryRowData partition, int bucket, List<DataFileMeta> files)
             throws IOException {
+        switch (writeMode) {
+            case APPEND_ONLY:
+                return createAppendOnlyReader(partition, bucket, files);
+
+            case CHANGE_LOG:
+                return createLSMReader(partition, bucket, files);
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreReadImpl.java:97]: NIT: `createMergeTreeReader`

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreReadImpl.java [hunk 7]
```diff
@@ -83,6 +89,34 @@ public FileStoreRead withValueProjection(int[][] projectedFields) {
...
+
+            default:
+                throw new UnsupportedOperationException("Unknown write mode: " + writeMode);
+        }
+    }
+
+    private RecordReader createAppendOnlyReader(
+            BinaryRowData partition, int bucket, List<DataFileMeta> files) throws IOException {
+        Preconditions.checkArgument(
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreReadImpl.java:106]: We can assert in `withDropDelete`.

**Comment by JingsongLi** [comment]
> For append-only table, the assignment policy is determined by the SinkRecordConverter#bucket, which will just calculate the hash code of the whole row to get the correct bucket. I think it it the correct direction to go.

We can create a JIRA to track this too, I don't know if the shuffle is needed.

File: flink-table-store-connector/src/test/java/org/apache/flink/table/store/connector/FileStoreTableITCase.java [hunk 8]
```diff
@@ -58,4 +63,18 @@ private void prepareEnv(TableEnvironment env, String path) {
     }
 
     protected abstract List<String> ddl();
+
+    protected List<Row> sql(String query, Object... args) {
+        return sql(true, query, args);
+    }
+
+    protected List<Row> sql(boolean batch, String query, Object... args) {
```
[the remark this answers is not in the document]
Comment (openinx) [reply flink-table-store-connector/src/test/java/org/apache/flink/table/store/connector/FileStoreTableITCase.java:71]: Let's just name it `batchSql`, the prefix `execute` is a relatively obvious semantics, and executing sql is a quite frequent method.  Let's use `batchSql` to keep this name a bit shorter.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/WriteMode.java (hunk 4 above)
Comment (openinx) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/WriteMode.java:26]: The correct sentence is : `Neither data deduplication nor any primary key constraints will be done when inserting rows into table store.`

**Comment by openinx** [comment]
Thanks all the comments from @JingsongLi , I've just updated the PR. Thanks.
